mm2src/coins/z_coin.rs

```diff
@@ -1421,40 +1463,53 @@ impl SwapOps for ZCoin {
             None => H0,
         };
 
+        let mut fee_output_valid = false;
+        let mut burn_output_valid = false;
         for shielded_out in z_tx.shielded_outputs.iter() {
-            if let Some((note, address, memo)) =
-                try_sapling_output_recovery(self.consensus_params_ref(), block_height, &DEX_FEE_OVK, shielded_out)
+            if self
+                .validate_dex_fee_output(
+                    shielded_out,
+                    &DEX_FEE_OVK,
+                    &self.z_fields.dex_fee_addr,
+                    block_height,
+                    fee_amount_sat,
+                    &expected_memo,
+                )
+                .map_err(|err| {
+                    MmError::new(ValidatePaymentError::WrongPaymentTx(format!(
+                        "Bad dex fee output: {}",
+                        err
+                    )))
+                })?
             {
-                if address != self.z_fields.dex_fee_addr {
-                    let encoded = encode_payment_address(z_mainnet_constants::HRP_SAPLING_PAYMENT_ADDRESS, &address);
-                    let expected = encode_payment_address(
-                        z_mainnet_constants::HRP_SAPLING_PAYMENT_ADDRESS,
-                        &self.z_fields.dex_fee_addr,
-                    );
-                    return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
-                        "Dex fee was sent to the invalid address {}, expected {}",
-                        encoded, expected
-                    )));
-                }
-
-                if note.value != amount_sat {
-                    return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
-                        "Dex fee has invalid amount {}, expected {}",
-                        note.value, amount_sat
-                    )));
-                }
-
-                if memo != expected_memo {
-                    return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
-                        "Dex fee has invalid memo {:?}, expected {:?}",
-                        memo, expected_memo
-                    )));
+                fee_output_valid = true;
+            }
+            if let Some(burn_amount_sat) = burn_amount_sat {
+                if self
+                    .validate_dex_fee_output(
+                        shielded_out,
+                        &DEX_FEE_OVK,
+                        &self.z_fields.dex_burn_addr,
+                        block_height,
+                        burn_amount_sat,
+                        &expected_memo,
+                    )
+                    .map_err(|err| {
+                        MmError::new(ValidatePaymentError::WrongPaymentTx(format!(
+                            "Bad burn output: {}",
+                            err
+                        )))
+                    })?
+                {
+                    burn_output_valid = true;
```

Comment:
I think `if` statement is not need just assign directly e.g 
```rust
                fee_output_valid = self
                .validate_dex_fee_output(
                    shielded_out,
                    &DEX_FEE_OVK,
                    &self.z_fields.dex_fee_addr,
                    block_height,
                    fee_amount_sat,
                    &expected_memo,
                )
                .map_err(|err| {
                    MmError::new(ValidatePaymentError::WrongPaymentTx(format!(
                        "Bad dex fee output: {}",
                        err
                    )))
                })?;
 ```

same with `burn_output_valid`

Reply:
I intended to make this function unified to validate txns both with or w/o burn output.
If to do like that, on the next iteration fee_output_valid will be cleared (when checking the burn output)

Reply:
Sorry returned the loop back :-)
Forgot we can't assume zcoin outputs ordering - they are shuffled
@borngraced

Reply:
ah okay, sorry for the confusion.